pkg/attribute/rootBag.go

```diff
@@ -0,0 +1,224 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attribute
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+
+	mixerpb "istio.io/api/mixer/api/v1"
+)
+
+// The rootBag is always the base of a chain of bags. It's state is updated
+// only via an Attributes proto. This code is not protected by locks because
+// we know it is always mutated in a single-threaded context.
+type rootBag struct {
+	strings  map[string]string
+	int64s   map[string]int64
+	float64s map[string]float64
+	bools    map[string]bool
+	times    map[string]time.Time
+	bytes    map[string][]uint8
+}
+
+var rootBags = sync.Pool{
+	New: func() interface{} {
+		return &rootBag{
+			strings:  make(map[string]string),
+			int64s:   make(map[string]int64),
+			float64s: make(map[string]float64),
+			bools:    make(map[string]bool),
+			times:    make(map[string]time.Time),
+			bytes:    make(map[string][]uint8),
+		}
+	},
+}
+
+func getRootBag() *rootBag {
+	return rootBags.Get().(*rootBag)
+}
+
+func (rb *rootBag) Close() {
+	rb.reset()
+	rootBags.Put(rb)
+}
+
+func (rb *rootBag) String(name string) (string, bool) {
+	r, b := rb.strings[name]
+	return r, b
+}
+
+func (rb *rootBag) Int64(name string) (int64, bool) {
+	r, b := rb.int64s[name]
+	return r, b
+}
+
+func (rb *rootBag) Float64(name string) (float64, bool) {
+	r, b := rb.float64s[name]
+	return r, b
+}
+
+func (rb *rootBag) Bool(name string) (bool, bool) {
+	r, b := rb.bools[name]
+	return r, b
+}
+
+func (rb *rootBag) Time(name string) (time.Time, bool) {
+	r, b := rb.times[name]
+	return r, b
+}
+
+func (rb *rootBag) Bytes(name string) ([]uint8, bool) {
+	r, b := rb.bytes[name]
+	return r, b
+}
+
+func (rb *rootBag) reset() {
+	// my kingdom for a clear method on maps!
+
+	for k := range rb.strings {
+		delete(rb.strings, k)
+	}
+
+	for k := range rb.int64s {
+		delete(rb.int64s, k)
+	}
+
+	for k := range rb.float64s {
+		delete(rb.float64s, k)
+	}
+
+	for k := range rb.bools {
+		delete(rb.bools, k)
+	}
+
+	for k := range rb.times {
+		delete(rb.times, k)
+	}
+
+	for k := range rb.bytes {
+		delete(rb.bytes, k)
+	}
+}
+
+// Ensure that all dictionary indices are valid and that all values
+// are in range.
+//
+// Note that since we don't have the attribute schema, this doesn't validate
+// that a given attribute is being treated as the right type. That is, an
+// attribute called 'source.ip' which is of type IP_ADDRESS could be listed as
+// a string or an int, and we wouldn't catch it here.
+func checkPreconditions(dictionary dictionary, attrs *mixerpb.Attributes) error {
+	for k := range attrs.StringAttributes {
+		if _, present := dictionary[k]; !present {
+			return fmt.Errorf("attribute index %d is not defined in the current dictionary", k)
+		}
+	}
+
+	for k := range attrs.Int64Attributes {
```

Comment:
could we return https://github.com/hashicorp/go-multierror  here?
Would be good to see all errors at one time.